conv2d: require bias to have the same dtype as input and weight on cpu

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83686

Fixes https://github.com/pytorch/pytorch/issues/83505

BC-breaking message:
- Previously we only required input and weight to have the same dtype on cpu (when input is non-complex). After this change, the dtype of bias is now also expected to have the same dtype. This change was necessary to improve the error message for certain combinations of inputs. This behavior now also matches that of convolution on cuda.

<details>
<summary>
Old plan
</summary>
Previously convolution (at least for slow_conv2d) did not perform type promotion, i.e. the output of `conv(int, int, float)` is an int, and that leads to the autograd assert.

This PR adds type promotion handling at the `at::native::conv2d` (this is a composite) level. We also need to correct or remove many tests that assume that conv errors when input types are mixed

Pros:
- Doing type promotion at this level avoids the complex path from having any special handling for mixed dtypes, and can potentially speed up mixed dtype inputs to now dispatch to faster kernels which are only capable of handling floats.

Cons:
- Doing type promotion at this level has the risk of introducing extra overhead when we would've dispatched to a kernel capable of handle mixed type anyway. I don't know if any of these exist at all though - it is possible that inputs with any non-float arguments are dispatched to the slow path. 

If this approach is OK, we can proceed with the other convolutions as well:
</details>

soulitzer committed:
Fix type promotion for slow conv2d

[ghstack-poisoned]

soulitzer commented on aten/src/ATen/native/ConvolutionMM2d.cpp:
```diff
@@ -615,16 +615,25 @@ Tensor& slow_conv2d_forward_out_cpu(
+  auto self = self_.to(common_type);
+  auto weight = weight_.to(common_type);
+  auto output = at::empty({0}, self.options().dtype(common_type));
   at::native::slow_conv2d_forward_out_cpu(
```
Maybe the type promotion should be performed in the out variant instead. We'd need to find the common dtype twice though.

soulitzer commented on aten/src/ATen/native/ConvolutionMM2d.cpp:
```diff
@@ -634,6 +643,7 @@ Tensor slow_conv2d_forward_cpu(
       padding,
       output);
 
+
```
There are probably more cases where this is happening, but we should make sure this is the right approach first.

soulitzer committed:
Update on "Fix type promotion for slow conv2d"


Fixes https://github.com/pytorch/pytorch/issues/83505

[ghstack-poisoned]

soulitzer committed:
Update on "Fix type promotion for slow conv2d"


Fixes https://github.com/pytorch/pytorch/issues/83505

[ghstack-poisoned]

soulitzer committed:
Update on "Fix type promotion for slow conv2d"


Fixes https://github.com/pytorch/pytorch/issues/83505

[ghstack-poisoned]

soulitzer committed:
Update on "Fix type promotion for slow conv2d"


Fixes https://github.com/pytorch/pytorch/issues/83505

[ghstack-poisoned]

albanD reviewed (commented):
Will let @ngimel review this one to make sure this is the right way to type promote (and make sure it is consistent with the ref?)

soulitzer committed:
Update on "Fix type promotion for slow conv2d"


Fixes https://github.com/pytorch/pytorch/issues/83505

[ghstack-poisoned]

soulitzer committed:
Update on "Fix type promotion for conv2d"


Fixes https://github.com/pytorch/pytorch/issues/83505

Previously convolution (at least for slow_conv2d) did not perform type promotion, i.e. the output of `conv(int, int, float)` is an int, and that leads to the autograd assert.

This PR adds type promotion handling at the `at::native::conv2d` (this is a composite) level.

Pros:
- Doing type promotion at this level avoids the complex path from having any special handling for mixed dtypes, and can potentially speed up mixed dtype inputs to now dispatch to faster kernels which are only capable of handling floats.

Cons:
- Doing type promotion at this level has the risk of introducing extra overhead when we would've dispatched to a kernel capable of handle mixed type anyway. I don't know if any of these exist at all though - it is possible that inputs with any non-float arguments are dispatched to the slow path. 

If this approach is OK, we can proceed with the other convolutions as well:

[ghstack-poisoned]

soulitzer committed:
Update on "Fix type promotion for conv2d"


Fixes https://github.com/pytorch/pytorch/issues/83505

Previously convolution (at least for slow_conv2d) did not perform type promotion, i.e. the output of `conv(int, int, float)` is an int, and that leads to the autograd assert.

This PR adds type promotion handling at the `at::native::conv2d` (this is a composite) level.

Pros:
- Doing type promotion at this level avoids the complex path from having any special handling for mixed dtypes, and can potentially speed up mixed dtype inputs to now dispatch to faster kernels which are only capable of handling floats.

Cons:
- Doing type promotion at this level has the risk of introducing extra overhead when we would've dispatched to a kernel capable of handle mixed type anyway. I don't know if any of these exist at all though - it is possible that inputs with any non-float arguments are dispatched to the slow path. 

If this approach is OK, we can proceed with the other convolutions as well:

[ghstack-poisoned]

soulitzer committed:
Update on "Fix type promotion for conv2d"


Fixes https://github.com/pytorch/pytorch/issues/83505

Previously convolution (at least for slow_conv2d) did not perform type promotion, i.e. the output of `conv(int, int, float)` is an int, and that leads to the autograd assert.

This PR adds type promotion handling at the `at::native::conv2d` (this is a composite) level.

Pros:
- Doing type promotion at this level avoids the complex path from having any special handling for mixed dtypes, and can potentially speed up mixed dtype inputs to now dispatch to faster kernels which are only capable of handling floats.

Cons:
- Doing type promotion at this level has the risk of introducing extra overhead when we would've dispatched to a kernel capable of handle mixed type anyway. I don't know if any of these exist at all though - it is possible that inputs with any non-float arguments are dispatched to the slow path. 

If this approach is OK, we can proceed with the other convolutions as well:

[ghstack-poisoned]

soulitzer committed:
Update on "Fix type promotion for conv2d"


Fixes https://github.com/pytorch/pytorch/issues/83505

Previously convolution (at least for slow_conv2d) did not perform type promotion, i.e. the output of `conv(int, int, float)` is an int, and that leads to the autograd assert.

This PR adds type promotion handling at the `at::native::conv2d` (this is a composite) level.

Pros:
- Doing type promotion at this level avoids the complex path from having any special handling for mixed dtypes, and can potentially speed up mixed dtype inputs to now dispatch to faster kernels which are only capable of handling floats.

Cons:
- Doing type promotion at this level has the risk of introducing extra overhead when we would've dispatched to a kernel capable of handle mixed type anyway. I don't know if any of these exist at all though - it is possible that inputs with any non-float arguments are dispatched to the slow path. 

If this approach is OK, we can proceed with the other convolutions as well:

[ghstack-poisoned]

soulitzer committed:
Update on "Fix type promotion for conv2d"


Fixes https://github.com/pytorch/pytorch/issues/83505

Previously convolution (at least for slow_conv2d) did not perform type promotion, i.e. the output of `conv(int, int, float)` is an int, and that leads to the autograd assert.

This PR adds type promotion handling at the `at::native::conv2d` (this is a composite) level. We also need to correct or remove many tests that assume that conv errors when input types are mixed

Pros:
- Doing type promotion at this level avoids the complex path from having any special handling for mixed dtypes, and can potentially speed up mixed dtype inputs to now dispatch to faster kernels which are only capable of handling floats.

Cons:
- Doing type promotion at this level has the risk of introducing extra overhead when we would've dispatched to a kernel capable of handle mixed type anyway. I don't know if any of these exist at all though - it is possible that inputs with any non-float arguments are dispatched to the slow path. 

If this approach is OK, we can proceed with the other convolutions as well:

[ghstack-poisoned]

soulitzer committed:
Update on "Fix type promotion for conv2d"


Fixes https://github.com/pytorch/pytorch/issues/83505

Previously convolution (at least for slow_conv2d) did not perform type promotion, i.e. the output of `conv(int, int, float)` is an int, and that leads to the autograd assert.

This PR adds type promotion handling at the `at::native::conv2d` (this is a composite) level. We also need to correct or remove many tests that assume that conv errors when input types are mixed

Pros:
- Doing type promotion at this level avoids the complex path from having any special handling for mixed dtypes, and can potentially speed up mixed dtype inputs to now dispatch to faster kernels which are only capable of handling floats.

Cons:
- Doing type promotion at this level has the risk of introducing extra overhead when we would've dispatched to a kernel capable of handle mixed type anyway. I don't know if any of these exist at all though - it is possible that inputs with any non-float arguments are dispatched to the slow path. 

If this approach is OK, we can proceed with the other convolutions as well:

[ghstack-poisoned]

soulitzer committed:
Update on "Fix type promotion for conv2d"


Fixes https://github.com/pytorch/pytorch/issues/83505

Previously convolution (at least for slow_conv2d) did not perform type promotion, i.e. the output of `conv(int, int, float)` is an int, and that leads to the autograd assert.

This PR adds type promotion handling at the `at::native::conv2d` (this is a composite) level. We also need to correct or remove many tests that assume that conv errors when input types are mixed

Pros:
- Doing type promotion at this level avoids the complex path from having any special handling for mixed dtypes, and can potentially speed up mixed dtype inputs to now dispatch to faster kernels which are only capable of handling floats.

Cons:
- Doing type promotion at this level has the risk of introducing extra overhead when we would've dispatched to a kernel capable of handle mixed type anyway. I don't know if any of these exist at all though - it is possible that inputs with any non-float arguments are dispatched to the slow path. 

If this approach is OK, we can proceed with the other convolutions as well:

[ghstack-poisoned]

ngimel commented:
I'm uneasy about letting convolution type promote, seems like it would enable more errors than it would solve. 
Re: cons, we don't have any kernels that are capable of handling mixed types, but errors with e.g. mixed float16/float32 inputs have been helpful in debugging errors in user script, this way they would just be silently promoted to fp32.

soulitzer commented:
If we don't want to do type promotion here, perhaps the alternative is to explicitly error (at the composite level before reaching autograd). The main risk of that seems to be that it would be bc-breaking.

Currently convolution kernel on cpu only checks whether input and weight have the same dtype (I only checked the slow_conv2d kernel). So combinations like the following work:
- `int, int, float` (only when the bias does not require grad, or else we run into the autograd issue)
- `float, float, complex` and `int, int, complex` (this is wrong because we end up in-place copying a complex into non-complex, but we do correctly warn that we are casting away the imag component) 
- `float, float, int`
- `floatA, floatA, floatB`

Since (X, X, `complex`) is wrong anyway (I don't know how useful the other cases are either), perhaps we are okay with making the bc-breaking change to make sure the dtype of bias matches that of input. This also matches behavior on cuda.

Alternatively, to avoid any bc-breakness but still get rid of the autograd internal assert we could condition erroring for mixed dtype inputs under the following condition:
- input matches the `int, int, float/complex` case AND
  -  (bias `requires_grad=True` AND grad_mode enabled AND `inference mode` not enabled)

The second approach seems very hacky and makes the behavior of convolution for mixed dtypes even more complex, so I'd probably prefer the first one though I'm not completely sure how severe the bc-breakingness is.

ngimel commented:
Let's try bc-breaking and check bias dtype to match input and weight, since it already matches cuda I hope that should be fine.

soulitzer committed:
Update on "Fix type promotion for conv2d"


Fixes https://github.com/pytorch/pytorch/issues/83505

Previously convolution (at least for slow_conv2d) did not perform type promotion, i.e. the output of `conv(int, int, float)` is an int, and that leads to the autograd assert.

This PR adds type promotion handling at the `at::native::conv2d` (this is a composite) level. We also need to correct or remove many tests that assume that conv errors when input types are mixed

Pros:
- Doing type promotion at this level avoids the complex path from having any special handling for mixed dtypes, and can potentially speed up mixed dtype inputs to now dispatch to faster kernels which are only capable of handling floats.

Cons:
- Doing type promotion at this level has the risk of introducing extra overhead when we would've dispatched to a kernel capable of handle mixed type anyway. I don't know if any of these exist at all though - it is possible that inputs with any non-float arguments are dispatched to the slow path. 

If this approach is OK, we can proceed with the other convolutions as well:

[ghstack-poisoned]

soulitzer committed:
Update on "Fix type promotion for conv2d"


Fixes https://github.com/pytorch/pytorch/issues/83505

Previously convolution (at least for slow_conv2d) did not perform type promotion, i.e. the output of `conv(int, int, float)` is an int, and that leads to the autograd assert.

This PR adds type promotion handling at the `at::native::conv2d` (this is a composite) level. We also need to correct or remove many tests that assume that conv errors when input types are mixed

Pros:
- Doing type promotion at this level avoids the complex path from having any special handling for mixed dtypes, and can potentially speed up mixed dtype inputs to now dispatch to faster kernels which are only capable of handling floats.

Cons:
- Doing type promotion at this level has the risk of introducing extra overhead when we would've dispatched to a kernel capable of handle mixed type anyway. I don't know if any of these exist at all though - it is possible that inputs with any non-float arguments are dispatched to the slow path. 

If this approach is OK, we can proceed with the other convolutions as well:

[ghstack-poisoned]

soulitzer commented:
@ngimel, this should be ready for another look!

ngimel reviewed (approved):
Looks great, thanks!

soulitzer commented:
@pytorchbot merge -g
